File: umbrella/ios/src/commonMain/kotlin/dev/valeryvpetrov/decision/umbrella/ios/FlowExt.kt
package dev.valeryvpetrov.decision.umbrella.ios

import kotlinx.coroutines.Job
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onCompletion
import kotlinx.coroutines.flow.onEach

fun Flow<*>.subscribe(
    onEach: (item: Any?) -> Unit,
    onComplete: () -> Unit,
    onThrow: (error: Throwable) -> Unit,
): Job =
    this.onEach { onEach(it) }
        .catch { onThrow(it) }
        .onCompletion { onComplete() }
        .launchIn(MainScope())

File: app/iosNative/iosApp/KotlinCoroutinesHelpers/FlowExt.swift
//
//  FlowExt.swift
//  iosApp
//
//  Created by Валерий Петров on 28.08.2024.
//  Copyright © 2024 orgName. All rights reserved.
//

import Combine
import UmbrellaIos

@available(iOS 13.0, *)
struct FlowPublisher<T: Any>: Publisher {

    public typealias Output = T
    public typealias Failure = Never
    private let flow: Flow

    public init(flow: Flow) {
        self.flow = flow
    }

    public func receive<S: Subscriber>(subscriber: S) where S.Input == T, S.Failure == Failure {
        subscriber.receive(subscription: FlowSubscription(flow: flow, subscriber: subscriber))
    }

    final class FlowSubscription<S: Subscriber>: Subscription where S.Input == T, S.Failure == Failure {

        private var subscriber: S?
        private var job: Job?
        private let flow: Flow

        init(flow: Flow, subscriber: S) {
            self.flow = flow
            self.subscriber = subscriber
        }

        func request(_ demand: Combine.Subscribers.Demand) {

            guard let subscriber = subscriber else {
                return
            }

            job = FlowExtKt.subscribe(
                flow,
                onEach: { item in if let item = item as? T { _ = subscriber.receive(item) } },
                onComplete: { subscriber.receive(completion: .finished) },
                onThrow: { error in debugPrint(error) }
            )
        }

        func cancel() {
            subscriber = nil
            job?.cancel(cause: nil)
        }
    }
}

@available(iOS 13.0, *)
public extension Flow {
    func asPublisher<T: AnyObject>() -> AnyPublisher<T, Never> {
        (FlowPublisher(flow: self) as FlowPublisher<T>).eraseToAnyPublisher()
    }
}
